### PR TITLE
KAFKA-15460: Add group type filter to List Groups API

### PR DESCRIPTION
**Pull request by rreddy-22**
In KIP-848, we introduce the notion of Group Types based on the protocol type that the members in the consumer group use. As of now we support two types of groups:

Classic : Members use the classic consumer group protocol ( existing one )
Consumer : Members use the consumer group protocol introduced in KIP-848.
Currently List Groups allows users to list all the consumer groups available. KIP-518 introduced filtering the consumer groups by the state that they are in. We now want to allow users to filter consumer groups by type.

This PR includes all the changes to the API and the listGroups implementation in both the old group coordinator and the new group coordinator.

### CODE REVIEW

File: clients/src/main/resources/common/message/ListGroupsResponse.json
Comment (dajac) [inline clients/src/main/resources/common/message/ListGroupsResponse.json]: nit: Let's revert this change.

File: clients/src/main/resources/common/message/ListGroupsResponse.json
Comment (dajac) [inline clients/src/main/resources/common/message/ListGroupsResponse.json]: nit: Let's revert this change.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: In my opinion, we should rather consider all groups as `classic` groups. Therefore, we would only return groups if the type is empty or contains `classic`.

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: Let's revert this change.

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala
Comment (dajac) [inline core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:332]: Should we also add variants with group types?

File: clients/src/main/java/org/apache/kafka/common/ConsumerGroupType.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/common/ConsumerGroupType.java]: This seems to be client side related. Do we need it in this PR?

File: clients/src/main/java/org/apache/kafka/common/requests/ListGroupsRequest.java
Comment (dajac) [inline clients/src/main/java/org/apache/kafka/common/requests/ListGroupsRequest.java:56]: Should we add a unit test to cover this change?

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: Most of the changes in this file are not related to this PR. Could we revert all the unnecessary changes please?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: `parallelStream()` is interesting here. I would start without it and measure the impact when we run benchmarks.

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the PR. I left some comments. Could you also check the tests? I think that there are some related failures. I would also be great if we could extend the integration tests for this API.

File: clients/src/main/java/org/apache/kafka/common/ConsumerGroupType.java
Comment (rreddy-22) [reply clients/src/main/java/org/apache/kafka/common/ConsumerGroupType.java]: Yes it is, we can consider it part of the admin client PR but there was a dependency in this one too

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (rreddy-22) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Thanks that makes sense, I'll make the changes, was waiting on your comments on my design doc to make the final call

File: core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala
Comment (rreddy-22) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:332]: Yep thanks for bringing it up, we could add it, I considered it but it didn't seem to add much value to the test so I decided against it at the time.

Comment (rreddy-22) [reply core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:332]: Yep thanks for mentioning this, let me add it in, at the time I didn't think it was necessary but I can add it now

File: core/src/test/scala/unit/kafka/server/KafkaApisTest.scala
Comment (rreddy-22) [reply core/src/test/scala/unit/kafka/server/KafkaApisTest.scala]: yep sorry about that, when I rebased and transferred the files  I might've messed up

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: thanks for the comment! okay I can make the change :) My understanding was that we want to scale to a large number of groups so I was trying to optimize wherever I could!

File: clients/src/main/java/org/apache/kafka/common/requests/ListGroupsRequest.java
Comment (rreddy-22) [reply clients/src/main/java/org/apache/kafka/common/requests/ListGroupsRequest.java:56]: Oh yes! Thanks for noticing this, I missed this test!

File: clients/src/main/java/org/apache/kafka/common/ConsumerGroupType.java
Comment (dajac) [reply clients/src/main/java/org/apache/kafka/common/ConsumerGroupType.java]: It seems that we only use in tests. Could we use `org.apache.kafka.coordinator.group.Group.GroupType` enum instead?

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: nit: Could we use `GroupType.CLASSIC` instead of hardcoding `classic` here?

File: core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala]: nit: ditto about using GroupType.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: Should we use Set for those two?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: How do we handle the case insensitive?

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:9591]: I don't see any test cases with group types. Should we add some?

**Review by dajac — COMMENTED** [review]
@rreddy-22 I made another pass and I left a few more comments.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:472]: I didn't implement it by then, was still waiting for confirmation from today's meeting, It is there now!

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I decided to convert the the arg to lower case in the handle list groups request itself, so we wouldn't need to check for case in any of the API portions

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (rreddy-22) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: Yes okie I'll do that thanks!

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java
Comment (rreddy-22) [reply group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:9591]: I think it's there, did you want a different test by any chance?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Sure I can do that, I just left it like how it was before

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: From the ListGroupsDataRequest the filters are passed as lists though, did you want to just convert them to sets here or throughout the whole pipeline?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: as discussed offline, we cannot rely on the client side to do the right thing. we should rather lower case them when on he server side.

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Makes sense, I changed it back

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: It would be better to do this outside of the predicate to avoid doing it for every group.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:9679]: Should we also test with the other type?

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I left a few minor comments.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: oya that makes sense, thanks!

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java
Comment (rreddy-22) [reply group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:9679]: Okie I added the case insensitive case to be of Consumer type

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: I wonder if it is worth storing the lower case type in the GroupType. What do you think? An alternative would be to convert the String to GroupType at L470 and compare enums here.

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I left one more comment.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I also considered just adding the lower case group types to the enum itself but I haven't seen that before and was also confused if it would make sense. What do you think?

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: For the second option my understanding is that to convert the string to groupType we would still need to convert it to lower case to compare and see if it is a valid group type. Lmk if I'm missing something more obvious

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I was also under the impression that case insensitive meant that as long as the word spells the correct type we don't care what the cases of the letters are? So currently if someone were to type "cLaSsic" that would still be valid. If we want to strictly allow only "Classic" and "classic" we would need to probably add both for each type to the enum itself. Lmk what you think?

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Sorry, I was not clear. I do think that we need to lowercase somewhere. However, I was trying to avoid doing it here while iterating over all the groups. This is a waste. This is why I suggested to store the lowercase version of the type in the enum or convert to the enum before. Do you see what I mean?

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I think I understand what you're saying. I'll add a way to parse any string and convert it to the necessary case to compare and determine which type it is in the group type enum. However, since some of these changes are on the client side can we do it in the second PR? That way I can test it as well. I could remove the lower case conversion here and then directly handle it there or we could change it there directly as an optimization just to be safe. What do you think?

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: None of the required changes are on the client side or I don't see them. I would prefer to address it in this PR.

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Discussed offline. We realized we were talking about two different GroupType files. Made the changes to the server side GroupType and added string to enum conversion and comparison

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:480]: We need to discuss the handling of unknown types. I thought that we would ignore the unknown types to basically follow what we did for the states. I think that you're suggesting to throw an exception for them. I think that the issue with throwing an exception is that the end user won't really know what is going on because we don't have an error message in the response. He would get an unknown server error. Therefore, I mean towards ignoring the unknown types. What do you think?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java]: Is there a reason why you did not build a Map to do the reverse lookup like you did in https://github.com/apache/kafka/pull/15259. By the way, I was also considering whether we should reuse the client GroupType on the server. What do you think?

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I left a few comments.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java]: We don't have to change to using the client side type is this PR though.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:480]: Yeah I was actually unsure about this as well. 
In the consumerGroupCommand tool we return an illegalArgumentException if the type is unknown so I thought maybe it would make sense to do that here as well. 
To ignore it should parse return null? If it returns null I was worried about NullPointerExceptions wherever this method is used. If I wanted to return UNKNOWN type like how it is on the client side, we would have to add that to the enum on the server side. Is there some obvious way to do this that I'm missing?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java]: Haha honestly there's no good reason, I used the existing implementation for states as inspiration in the client side one and I just implemented it differently here. I can use the reverse lookup here too.

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java]: Yes we could use the same enum on the server too, should we change it in the other PR?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:480]: We were able to ignore the unknown states easily because we didn't have to parse them and assign an enum type right? It would just check if the state exists in the filter set, but here since we're tryna parse and assign ig it needs to be handled differently? Lmk if my understanding is off

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:480]: > In the consumerGroupCommand tool we return an illegalArgumentException if the type is unknown so I thought maybe it would make sense to do that here as well.

Yeah, the context is a bit different here. You have to think about how illegalArgumentException will get returned to the user. The Kafka protocol does not have any error for this so it converts it to a Unknown Server Error without any details. This is not great from a user perspective.

I think that the correct way would be to try parsing the received group type. If it succeeds, great, we keep it. Otherwise, we simply discard it. If you want to keep the illegalArgumentException, you could catch it when you call `parse`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/Group.java]: Yeah, we don't have to change it here as it is not related. The minimum to do here is to also build a Map within the enum to do the lookup. It is better to be consistent.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:480]: I initially threw the illegalArgException and then caught it and ignored it in the listGroups method but this caused the opposite result from what we wanted aka it returns all the groups since the typeFilter is empty. The only way to get the expected behavior is to add unknown to the group type enum. This gives us the same result as when an unknown state filter is passed. I even added tests to make sure. Lmk if this doesn't work or if there was something else you had in mind.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: nit: Could we bring back `caseInsensitiveStates` in order to not lower case the states for every group?

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I left one small nit. We should be good to go when addressed.

File: core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala
Comment (dajac) [inline core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: nit: Could we put `()` around `states.isEmpty || g.isInStates(caseInsensitiveStates)` to make the grouping clear?

Comment (rreddy-22) [reply core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala]: sure :)

**Comment by dajac** [comment]
@rreddy-22 Could you please check the build? It looks like there is a failure related to checkstyle.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Could we say: `Loaded group {} with an unknown type {}.`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: Could we say `Removed group {} with an unknown type {}`?

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I left two more nits.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Does it still load the group if the type is unknown though?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Does it still remove the group?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: The group is in `groups` so it is already loaded at this point.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Looking at the first instruction in this method, the group is removed from `groups`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Oh yeah my bad, morning brain :(

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: There is an extra space after the `.` and the `+` must be replaced by a `,`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: Same comment.

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I just found two small nits while reviewing it. Could you please take a look?

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Pull request merged**
